contracts/examples/erc1155/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+#![no_std]
+
+use elrond_codec::test_util::top_encode_to_vec_or_panic;
+use elrond_wasm::HexCallDataSerializer;
+
+elrond_wasm::imports!();
+elrond_wasm::derive_imports!();
+
+const INTERFACE_SIGNATURE_ERC165: u32 = 0x01ffc9a7;
+const INTERFACE_SIGNATURE_ERC1155: u32 = 0xd9b67a26;
+
+const ON_ERC_RECEIVED_ENDPOINT_NAME: &[u8] = b"onERC1155Received";
+const ON_ERC_BATCH_RECEIVED_ENDPOINT_NAME: &[u8] = b"onERC1155BatchReceived";
+const ACCEPTED_TRANSFER_ANSWER: u32 = 0xbc197c81;
+
+#[derive(TopEncode, TopDecode)]
+pub struct Transfer<BigUint: BigUintApi> {
+	pub from: Address,
+	pub to: Address,
+	pub type_ids: Vec<BigUint>,
+	pub values: Vec<BigUint>,
+}
+
+#[elrond_wasm_derive::contract(Erc1155Impl)]
+pub trait Erc1155 {
+	#[init]
+	fn init(&self) {}
+
+	// endpoints
+
+	/// `value` is amount for fungible, token_id for non-fungible
+	#[endpoint(safeTransferFrom)]
+	fn safe_transfer_from(
+		&self,
+		from: Address,
+		to: Address,
+		type_id: BigUint,
+		value: BigUint,
+		data: &[u8],
+	) -> SCResult<()> {
+		let caller = self.get_caller();
+
+		require!(to != Address::zero(), "Can't transfer to address zero");
+		require!(self.is_valid_type_id(&type_id), "Token id is invalid");
+		require!(
+			caller == from || self.get_is_approved(&caller, &from),
+			"Caller is not approved to transfer tokens from address"
+		);
+
+		if self.get_is_fungible(&type_id) == true {
+			let amount = &value;
+
+			sc_try!(self.try_reserve_fungible(&from, &type_id, &amount));
+
+			if self.is_smart_contract_address(&to) {
+				self.peform_async_call_single_transfer(from, to, type_id, value, data);
+			} else {
+				self.increase_balance(&to, &type_id, &amount);
+			}
+		} else {
+			let token_id = &value;
+
+			sc_try!(self.try_reserve_non_fungible(&from, &type_id, &token_id));
+
+			if self.is_smart_contract_address(&to) {
+				self.peform_async_call_single_transfer(from, to, type_id, value, data);
+			} else {
+				let amount = BigUint::from(1u32);
+				self.increase_balance(&to, &type_id, &amount);
+				self.set_token_owner(&type_id, token_id, &to);
+			}
+		}
+
+		// self.transfer_single_event(&caller, &from, &to, &id, &amount);
+
+		Ok(())
+	}
+
+	/// `value` is amount for fungible, token_id for non-fungible
+	#[endpoint(safeBatchTransferFrom)]
+	fn safe_batch_transfer_from(
+		&self,
+		from: Address,
+		to: Address,
+		type_ids: &[BigUint],
```

Comment:
Optional: if the 2 lists must always have the same length, maybe it would be better to provide that data as a single list of pairs

Reply:
I'll leave them like this for now, it's a bit clearer and more consistent with the transfer single endpoint, but maybe I'll change at some point. Good tip though!